ICSharpCode.Decompiler/CSharp/OutputVisitor/InsertParenthesesVisitor.cs

```diff
@@ -389,6 +390,36 @@ public override void VisitAsExpression(AsExpression asExpression)
 			base.VisitAsExpression(asExpression);
 		}
 
+		public override void VisitInterpolation(Interpolation interpolation)
+		{
+			// If an interpolation contains global::, we need to parenthesize the expression.
+			if (InterpolationNeedsParenthesized(interpolation))
+				Parenthesize(interpolation.Expression);
+			base.VisitInterpolation(interpolation);
+
+			static bool InterpolationNeedsParenthesized(AstNode node)
+			{
+				if (node is MemberType { IsDoubleColon: true })
+					return true;
+
+				if (node is ConditionalExpression)
+					return false; // Conditional expressions are parenthesized in their own visit method.
+				if (node is AnonymousMethodExpression or LambdaExpression)
+					return false;
+				if (node is InvocationExpression invocation)
+					return InterpolationNeedsParenthesized(invocation.Target);
+				if (node is CastExpression cast)
+					return InterpolationNeedsParenthesized(cast.Expression);
```

Comment:
There is a whole lot of special cases for cast-expressions: see `InsertParenthesesVisitor.VisitCastExpression`.

Reply:
I reordered the visitation of children, so that those special cases should get hit before we reason about whether or not to parenthesize the interpolation expression.